File: framework/java/src/main/java/soda/unittest/validate/UnorderedListFeature.java
package soda.unittest.validate;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class UnorderedListFeature<T> implements ObjectFeature<List<T>> {

    private ObjectFeature<T> elemFeat;

    public UnorderedListFeature(ObjectFeature<T> elemFeat) {
        this.elemFeat = elemFeat;
    }

    @Override
    public long hash(List<T> obj) {
        var hashes = new ArrayList<Long>();
        for (var e : obj) {
            hashes.add(elemFeat.hash(e));
        }
        Collections.sort(hashes);
        long res = 0;
        for (long h : hashes) {
            res = res * 133 + h;
        }
        return res;
    }

    @Override
    public boolean isEqual(List<T> a, List<T> b) {
        if (a.size() != b.size()) {
            return false;
        }
        var xmap = new XMap<T,Integer>(elemFeat);
        for (var e : a) {
            xmap.put(e, xmap.getOrDefault(e, 0) + 1);
        }
        for (var e : b) {
            if (!xmap.contains(e)) {
                return false;
            }
            int c = xmap.get(e) - 1;
            if (c == 0) {
                xmap.remove(e);
            } else {
                xmap.put(e, c);
            }
        }
        return true;
    }
}

File: framework/scala/src/main/scala/soda/scala/unittest/validate/UnorderedListFeature.scala
package soda.scala.unittest.validate

class UnorderedListFeature[T](elemFeat: ObjectFeature[T]) extends ObjectFeature[List[T]] {

  override def hash(obj: List[T]): Long = {
    obj.map(elemFeat.hash).sorted.foldLeft(0L)((a, b) => a * 133 + b)
  }

  override def isEqual(a: List[T], b: List[T]): Boolean = {
    if (a.size != b.size) {
      return false
    }
    val xmap = new XMap[T, Int](elemFeat)
    for (e <- a) {
      xmap.put(e, xmap.getOrElse(e, 0) + 1)
    }
    for (e <- b) {
      if (!xmap.contains(e)) {
        return false
      }
      val c = xmap.get(e).get - 1
      if (c == 0) {
        xmap.remove(e)
      } else {
        xmap.put(e, c)
      }
    }
    true
  }

}
